File: proto/src/Rue/Proto/Verdict.hs
{-# LANGUAGE OverloadedStrings #-}
-- | The verdict's structured form, docs/ROADMAP.md section 5.8, and its
-- canonical JSON. The prose is Rue.Proto.Prose; @explain@ is
-- Rue.Proto.Explain. Every "the verdict says" in the roadmap names a field
-- here; a new clause without a field is a schema bump.
module Rue.Proto.Verdict
  ( Verdict (..)
  , Status (..)
  , Diagnostic (..)
  , PointOfNoReturn (..)
  , GateVerdict (..)
  , BackstopVerdict (..)
  , HostTouched (..)
  , StepVerdict (..)
  , StepGateVerdict (..)
  , KnellVerdict (..)
  , Dispatch (..)
  , MayConflict (..)
  , verdictVersion
  , toJson
  ) where

import Data.Aeson (Value (..), object, (.=))
import qualified Data.Aeson.Key as Key
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Vector as V
import Rue.Proto.Diagnostics (Code, codeText)
import Rue.Proto.Intent (Intent (..))
import Rue.Proto.Model (Duration (..))

verdictVersion :: Int
verdictVersion = 1

data Status = Ok | RefusedStatus
  deriving (Eq, Show)

data Diagnostic = Diagnostic
  { diagCode :: Code
  , diagStep :: Maybe Int
  , diagMessage :: Text
  }
  deriving (Eq, Show)

data PointOfNoReturn = PointOfNoReturn
  { ponrStep :: Int
  , ponrGuard :: Maybe Text
  , ponrCost :: Text
  , ponrAck :: Text
  , ponrGate :: Maybe Text
  }
  deriving (Eq, Show)

data GateVerdict = GateVerdict
  { gvSatisfiable :: Bool
  , gvMinDistinctHumans :: Maybe Int
  , gvZeroHumanPath :: Bool
  , gvWindow :: Maybe Duration
  , gvWaitAloneAt :: Maybe Duration
  }
  deriving (Eq, Show)

data BackstopVerdict = BackstopVerdict
  { bvTriggers :: [Value]
  , bvCovers :: [Int]
  , bvInstalledBefore :: Maybe Int
  , bvArmedBefore :: Maybe Int -- ^ armed before this step, or
  , bvArmedAfter :: Maybe Int -- ^ armed after this step (late arming)
  , bvLateArmingWindow :: [Int]
  , bvScheduler :: Text
  , bvGranularity :: Duration
  , bvSelfEnforced :: Bool
  , bvDriftPolicy :: [(Int, Text)]
  , bvSnapshotLocation :: Text
  , bvSnapshotCap :: Int
  , bvConditional :: [(Int, Text)]
  }
  deriving (Eq, Show)

data HostTouched
  = HostTouched Text Text -- ^ host, directory (@target@ or @controller@)
  | HostUnresolved Text -- ^ where the binding comes from
  deriving (Eq, Show)

data Dispatch = Dispatch {dispatchSource :: Text, dispatchHostContractHash :: Text}
  deriving (Eq, Show)

data MayConflict = MayConflict {mcEarlier :: Int, mcLater :: Int, mcFact :: Text, mcRefused :: Bool}
  deriving (Eq, Show)

data StepGateVerdict = StepGateVerdict
  { sgExpr :: Text
  , sgWindow :: Maybe Duration
  , sgWaitAloneAt :: Maybe Duration
  }
  deriving (Eq, Show)

data KnellVerdict = KnellVerdict
  { kvGuard :: Maybe Text
  , kvCost :: Text
  , kvAck :: Text
  }
  deriving (Eq, Show)

data StepVerdict = StepVerdict
  { svN :: Int
  , svOp :: Text
  , svLocus :: Value
  , svUndo :: Maybe Text
  , svUndoLocus :: Text
  , svRefusal :: Text
  , svDrift :: Maybe Text
  , svGate :: Maybe StepGateVerdict
  , svKnell :: Maybe KnellVerdict
  , svConditional :: Maybe Text
  , svFootprint :: [Text]
  }
  deriving (Eq, Show)

data Verdict = Verdict
  { vPlan :: Text
  , vHost :: Text
  , vStatus :: Status
  , vIntent :: Maybe Intent
  , vRehearsal :: Bool
  , vCommitStep :: Maybe Int
  , vFiresByConstruction :: Bool
  , vWane :: Maybe Duration
  , vReversibleThrough :: Int
  , vHoldsAt :: [Int]
  , vPointOfNoReturn :: Maybe PointOfNoReturn
  , vReversibleBackTo :: Maybe (Int, Int)
  , vGate :: Maybe GateVerdict
  , vBackstop :: Maybe BackstopVerdict
  , vControllerOnlyUndos :: [Int]
  , vHeldIndefinitely :: [Int]
  , vInducedDefer :: [Int]
  , vHostsTouched :: [(Int, [HostTouched])]
  , vDeferred :: [Int]
  , vDispatch :: Dispatch
  , vMayConflicts :: [MayConflict]
  , vUnresolvedBindings :: [Text]
  , vDiagnostics :: [Diagnostic]
  , vSteps :: [StepVerdict]
  }
  deriving (Eq, Show)

-- ---------------------------------------------------------------------------
-- JSON

stepKey :: Int -> Key.Key
stepKey = Key.fromText . T.pack . show

ints :: [Int] -> Value
ints = Array . V.fromList . map (Number . fromIntegral)

dur :: Maybe Duration -> Value
dur = maybe Null (\(Duration s) -> Number (fromIntegral s))

txt :: Maybe Text -> Value
txt = maybe Null String

toJson :: Verdict -> Value
toJson v =
  object
    [ "verdict_version" .= verdictVersion
    , "plan" .= vPlan v
    , "host" .= vHost v
    , "status" .= (case vStatus v of Ok -> "ok" :: Text; RefusedStatus -> "refused")
    , "intent" .= maybe Null (\i -> String (case i of Temporary -> "temporary"; Permanent -> "permanent")) (vIntent v)
    , "rehearsal" .= vRehearsal v
    , "commit_step" .= maybe Null (Number . fromIntegral) (vCommitStep v)
    , "fires_by_construction" .= vFiresByConstruction v
    , "wane_s" .= dur (vWane v)
    , "reversible_through" .= vReversibleThrough v
    , "holds_at" .= ints (vHoldsAt v)
    , "point_of_no_return" .= maybe Null ponr (vPointOfNoReturn v)
    , "reversible_back_to" .= maybe Null (\(f, t) -> object ["from" .= f, "to" .= t]) (vReversibleBackTo v)
    , "gate" .= maybe Null gate (vGate v)
    , "backstop" .= maybe Null backstop (vBackstop v)
    , "controller_only_undos" .= ints (vControllerOnlyUndos v)
    , "held_indefinitely" .= ints (vHeldIndefinitely v)
    , "induced_defer" .= ints (vInducedDefer v)
    , "hosts_touched" .= object [stepKey n .= Array (V.fromList (map host hs)) | (n, hs) <- vHostsTouched v]
    , "deferred" .= ints (vDeferred v)
    , "dispatch" .= object ["source" .= dispatchSource (vDispatch v), "host_contract_hash" .= dispatchHostContractHash (vDispatch v)]
    , "may_conflicts" .= Array (V.fromList (map mayConflict (vMayConflicts v)))
    , "unresolved_bindings" .= vUnresolvedBindings v
    , "diagnostics" .= Array (V.fromList (map diagnostic (vDiagnostics v)))
    , "steps" .= Array (V.fromList (map stepV (vSteps v)))
    ]
  where
    ponr p =
      object
        [ "step" .= ponrStep p
        , "guard" .= txt (ponrGuard p)
        , "cost" .= ponrCost p
        , "ack" .= ponrAck p
        , "gate" .= txt (ponrGate p)
        ]
    gate g =
      object
        [ "satisfiable" .= gvSatisfiable g
        , "min_distinct_humans" .= maybe Null (Number . fromIntegral) (gvMinDistinctHumans g)
        , "zero_human_path" .= gvZeroHumanPath g
        , "window_s" .= dur (gvWindow g)
        , "wait_alone_at_s" .= dur (gvWaitAloneAt g)
        ]
    backstop b =
      object
        [ "triggers" .= bvTriggers b
        , "covers" .= ints (bvCovers b)
        , "locus" .= ("target" :: Text)
        , "installed_before" .= maybe Null (Number . fromIntegral) (bvInstalledBefore b)
        , "armed_before" .= maybe Null (Number . fromIntegral) (bvArmedBefore b)
        , "armed_after" .= maybe Null (Number . fromIntegral) (bvArmedAfter b)
        , "late_arming_window" .= ints (bvLateArmingWindow b)
        , "scheduler" .= bvScheduler b
        , "granularity_s" .= seconds (bvGranularity b)
        , "self_enforced" .= bvSelfEnforced b
        , "drift_policy" .= object [stepKey n .= p | (n, p) <- bvDriftPolicy b]
        , "snapshots" .= object ["location" .= bvSnapshotLocation b, "cap_bytes" .= bvSnapshotCap b]
        , "conditional" .= Array (V.fromList [object ["step" .= n, "on" .= c] | (n, c) <- bvConditional b])
        ]
    host h = case h of
      HostTouched hn d -> object ["host" .= hn, "directory" .= d]
      HostUnresolved src -> object ["unresolved" .= src]
    mayConflict m =
      object ["earlier" .= mcEarlier m, "later" .= mcLater m, "fact" .= mcFact m, "refused" .= mcRefused m]
    diagnostic d =
      object ["code" .= codeText (diagCode d), "step" .= maybe Null (Number . fromIntegral) (diagStep d), "message" .= diagMessage d]
    stepV s =
      object
        [ "n" .= svN s
        , "op" .= svOp s
        , "locus" .= svLocus s
        , "undo" .= txt (svUndo s)
        , "undo_locus" .= svUndoLocus s
        , "refusal" .= svRefusal s
        , "drift" .= txt (svDrift s)
        , "gate" .= maybe Null (\g -> object ["expr" .= sgExpr g, "step_digest" .= True, "window_s" .= dur (sgWindow g), "wait_alone_at_s" .= dur (sgWaitAloneAt g)]) (svGate s)
        , "knell" .= maybe Null (\k -> object ["guard" .= txt (kvGuard k), "cost" .= kvCost k, "ack" .= kvAck k]) (svKnell s)
        , "conditional" .= txt (svConditional s)
        , "footprint" .= svFootprint s
        ]

File: core/src/verdict.rs
//! The verdict's structured form, docs/ROADMAP.md section 5.8, and its
//! canonical JSON. The prose is [`crate::prose`]; `explain` is
//! [`crate::explain`]. Every "the verdict says" in the roadmap names a field
//! here; a new clause without a field is a schema bump. The JSON is written
//! field by field, as the prototype's is, so the bytes are the specification's
//! and not a derive's.

use serde_json::{json, Map, Value};

use crate::diagnostics::Code;
use crate::intent::Intent;
use crate::model::Duration;

pub const VERDICT_VERSION: u32 = 1;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Status {
    Ok,
    Refused,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Diagnostic {
    pub code: Code,
    pub step: Option<u32>,
    pub message: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PointOfNoReturn {
    pub step: u32,
    pub guard: Option<String>,
    pub cost: String,
    pub ack: String,
    pub gate: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GateVerdict {
    pub satisfiable: bool,
    pub min_distinct_humans: Option<usize>,
    pub zero_human_path: bool,
    pub window: Option<Duration>,
    pub wait_alone_at: Option<Duration>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BackstopVerdict {
    pub triggers: Vec<String>,
    pub covers: Vec<u32>,
    pub installed_before: Option<u32>,
    /// Armed before this step, or
    pub armed_before: Option<u32>,
    /// armed after this step (late arming).
    pub armed_after: Option<u32>,
    pub late_arming_window: Vec<u32>,
    pub scheduler: String,
    pub granularity: Duration,
    pub self_enforced: bool,
    pub drift_policy: Vec<(u32, String)>,
    pub snapshot_location: String,
    pub snapshot_cap: u64,
    pub conditional: Vec<(u32, String)>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum HostTouched {
    /// A host and the directory its markers live in (`target` or `controller`).
    Host { host: String, directory: String },
    /// Where the binding comes from.
    Unresolved(String),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Dispatch {
    pub source: String,
    pub host_contract_hash: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MayConflict {
    pub earlier: u32,
    pub later: u32,
    pub fact: String,
    pub refused: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StepGateVerdict {
    pub expr: String,
    pub window: Option<Duration>,
    pub wait_alone_at: Option<Duration>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct KnellVerdict {
    pub guard: Option<String>,
    pub cost: String,
    pub ack: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StepVerdict {
    pub n: u32,
    pub op: String,
    pub locus: String,
    pub undo: Option<String>,
    pub undo_locus: String,
    pub refusal: String,
    pub drift: Option<String>,
    pub gate: Option<StepGateVerdict>,
    pub knell: Option<KnellVerdict>,
    pub conditional: Option<String>,
    pub footprint: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Verdict {
    pub plan: String,
    pub host: String,
    pub status: Status,
    pub intent: Option<Intent>,
    pub rehearsal: bool,
    /// `manual` or `auto`: which hold and acknowledgement rules applied.
    pub mode: String,
    pub commit_step: Option<u32>,
    pub fires_by_construction: bool,
    pub wane: Option<Duration>,
    pub reversible_through: u32,
    pub holds_at: Vec<u32>,
    pub point_of_no_return: Option<PointOfNoReturn>,
    pub reversible_back_to: Option<(u32, u32)>,
    pub gate: Option<GateVerdict>,
    pub backstop: Option<BackstopVerdict>,
    pub controller_only_undos: Vec<u32>,
    pub held_indefinitely: Vec<u32>,
    pub induced_defer: Vec<u32>,
    pub hosts_touched: Vec<(u32, Vec<HostTouched>)>,
    pub deferred: Vec<u32>,
    pub dispatch: Dispatch,
    pub may_conflicts: Vec<MayConflict>,
    pub unresolved_bindings: Vec<String>,
    pub diagnostics: Vec<Diagnostic>,
    pub steps: Vec<StepVerdict>,
}

fn dur(d: Option<Duration>) -> Value {
    match d {
        Some(d) => json!(d.seconds),
        None => Value::Null,
    }
}

fn txt(t: &Option<String>) -> Value {
    match t {
        Some(s) => Value::String(s.clone()),
        None => Value::Null,
    }
}

fn opt_u32(n: Option<u32>) -> Value {
    match n {
        Some(n) => json!(n),
        None => Value::Null,
    }
}

fn step_keyed(pairs: impl Iterator<Item = (u32, Value)>) -> Value {
    let mut m = Map::new();
    for (n, v) in pairs {
        m.insert(n.to_string(), v);
    }
    Value::Object(m)
}

pub fn to_json(v: &Verdict) -> Value {
    json!({
        "verdict_version": VERDICT_VERSION,
        "plan": v.plan,
        "host": v.host,
        "status": match v.status { Status::Ok => "ok", Status::Refused => "refused" },
        "intent": match v.intent { Some(Intent::Temporary) => json!("temporary"), Some(Intent::Permanent) => json!("permanent"), None => Value::Null },
        "rehearsal": v.rehearsal,
        "mode": v.mode,
        "commit_step": opt_u32(v.commit_step),
        "fires_by_construction": v.fires_by_construction,
        "wane_s": dur(v.wane),
        "reversible_through": v.reversible_through,
        "holds_at": v.holds_at,
        "point_of_no_return": match &v.point_of_no_return {
            Some(p) => json!({ "step": p.step, "guard": txt(&p.guard), "cost": p.cost, "ack": p.ack, "gate": txt(&p.gate) }),
            None => Value::Null,
        },
        "reversible_back_to": match v.reversible_back_to {
            Some((f, t)) => json!({ "from": f, "to": t }),
            None => Value::Null,
        },
        "gate": match &v.gate {
            Some(g) => json!({
                "satisfiable": g.satisfiable,
                "min_distinct_humans": match g.min_distinct_humans { Some(n) => json!(n), None => Value::Null },
                "zero_human_path": g.zero_human_path,
                "window_s": dur(g.window),
                "wait_alone_at_s": dur(g.wait_alone_at),
            }),
            None => Value::Null,
        },
        "backstop": match &v.backstop {
            Some(b) => json!({
                "triggers": b.triggers,
                "covers": b.covers,
                "locus": "target",
                "installed_before": opt_u32(b.installed_before),
                "armed_before": opt_u32(b.armed_before),
                "armed_after": opt_u32(b.armed_after),
                "late_arming_window": b.late_arming_window,
                "scheduler": b.scheduler,
                "granularity_s": b.granularity.seconds,
                "self_enforced": b.self_enforced,
                "drift_policy": step_keyed(b.drift_policy.iter().map(|(n, p)| (*n, Value::String(p.clone())))),
                "snapshots": { "location": b.snapshot_location, "cap_bytes": b.snapshot_cap },
                "conditional": b.conditional.iter().map(|(n, c)| json!({ "step": n, "on": c })).collect::<Vec<_>>(),
            }),
            None => Value::Null,
        },
        "controller_only_undos": v.controller_only_undos,
        "held_indefinitely": v.held_indefinitely,
        "induced_defer": v.induced_defer,
        "hosts_touched": step_keyed(v.hosts_touched.iter().map(|(n, hs)| {
            (*n, Value::Array(hs.iter().map(|h| match h {
                HostTouched::Host { host, directory } => json!({ "host": host, "directory": directory }),
                HostTouched::Unresolved(src) => json!({ "unresolved": src }),
            }).collect()))
        })),
        "deferred": v.deferred,
        "dispatch": { "source": v.dispatch.source, "host_contract_hash": v.dispatch.host_contract_hash },
        "may_conflicts": v.may_conflicts.iter().map(|m| json!({ "earlier": m.earlier, "later": m.later, "fact": m.fact, "refused": m.refused })).collect::<Vec<_>>(),
        "unresolved_bindings": v.unresolved_bindings,
        "diagnostics": v.diagnostics.iter().map(|d| json!({ "code": d.code.as_str(), "step": opt_u32(d.step), "message": d.message })).collect::<Vec<_>>(),
        "steps": v.steps.iter().map(|s| json!({
            "n": s.n,
            "op": s.op,
            "locus": s.locus,
            "undo": txt(&s.undo),
            "undo_locus": s.undo_locus,
            "refusal": s.refusal,
            "drift": txt(&s.drift),
            "gate": match &s.gate {
                Some(g) => json!({ "expr": g.expr, "step_digest": true, "window_s": dur(g.window), "wait_alone_at_s": dur(g.wait_alone_at) }),
                None => Value::Null,
            },
            "knell": match &s.knell {
                Some(k) => json!({ "guard": txt(&k.guard), "cost": k.cost, "ack": k.ack }),
                None => Value::Null,
            },
            "conditional": txt(&s.conditional),
            "footprint": s.footprint,
        })).collect::<Vec<_>>(),
    })
}
